### PR TITLE
SLURM_script that runs pretrain.py

### PR DESCRIPTION
**Pull request by AstroVasseL**

### CODE REVIEW

File: asmtransformers/scripts/launch-pretrain.sh [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    --nnodes "$NNODES"
+    --nproc-per-node "$GPUS_PER_NODE" 
+    --rdzv-backend "c10d" 
+    --rdzv-endpoint "$MASTER_ADDR":"$MASTER_PORT" 
+)
+
+TRAINIG_ARGS=(
+    --data "/data/dataset"
+    --tokenizer "/data/tokenizer"
+    --config "/app/asmtransformers/models/multilingual_asmbert/config_large.json"
```
Comment (akaIDIOT) [inline asmtransformers/scripts/launch-pretrain.sh]: I'm assuming that these paths correspond to some defaults that are defined elsewhere? Can we include such a pointer here as a comment or something?

**Review by akaIDIOT — COMMENTED** [review]
Not familiar with the SLURM / torchrun setup this targets, would it help to rope in someone who's more in touch with that? :)

File: asmtransformers/scripts/launch-pretrain.sh (hunk 1 above)
Comment (AstroVasseL) [reply asmtransformers/scripts/launch-pretrain.sh]: Yes! When creating the container, the data, tokenizer and config need bind on these locations.

File: asmtransformers/launch-pretrain.sh [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    --rdzv-endpoint "$MASTER_ADDR":"$MASTER_PORT" 
+)
+
+TRAINIG_ARGS=(
+    --data "/data/dataset"
+    --tokenizer "/data/tokenizer"
+    --config "/app/asmtransformers/models/multilingual_asmbert/config_large.json"
+    --epoch 19
+    --batch-size 192
+    --gradient-accumulation-steps 1
```
Comment (judithvdw) [inline asmtransformers/launch-pretrain.sh]: If we trained with a larger batch size on Gracie should we not train with some gradient accumulation here? To emulate the same effective batch size?

File: asmtransformers/scripts/launch-pretrain.sh [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    --epoch 19
+    --batch-size 192
+    --gradient-accumulation-steps 1
+    --mlm-prob 0.15
+    --save-steps 5000
+    --dataloader-num-workers 4
+    --bf16
+    --tf32
```
Comment (judithvdw) [inline asmtransformers/scripts/launch-pretrain.sh:22]: with the lines above, some of these are now the same as the default in peretrain.py but explicitly mentioned, where as some arguments are just left on default without mentioning. 

I'm assuming because it is the ones we actually want to experiment with, but might be good to have a comment on that in the script here.

Comment (judithvdw) [reply asmtransformers/scripts/launch-pretrain.sh:22]: Also, maybe good to in this pull request also add some notes about this shell script in the readme, what it is for and how to work with it (like having to edit the arguments here).

Most stuff related to that can be in the separate readme on surf/slurm, but some mention might be good.

File: asmtransformers/pretrain.Containerfile [hunk 4]
```diff
@@ -1,6 +1,6 @@
 FROM python:3.13
 ARG PIP_PYPI_URL=https://pypi.org/simple
-COPY --parents asmtransformers scripts pyproject.toml pdm.lock README.md /app
+COPY --parents asmtransformers scripts launch-pretrain.sh pyproject.toml pdm.lock README.md /app
```
Comment (judithvdw) [inline asmtransformers/pretrain.Containerfile]: I know I am the one who originally put launch-pretrain.sh in the root in stead of in scripts (because I was trying stuff out and wanted to make sure it worked from the root without thinking about working directories). But not actually sure if it is necessary to have it there. @akaIDIOT usually has opinions about these things, I don't mind ;)

**Review by judithvdw — APPROVED** [review]
Looks good. Do you think, as we need to have this here anyways, we might as well put the other job script in the repository here too?

File: asmtransformers/scripts/launch-pretrain.sh (hunk 3 above)
Comment (judithvdw) [reply asmtransformers/scripts/launch-pretrain.sh:22]: One other thing I was wondering, is there a way to pass these on from the script outside the container? Because now every time you want to do a different experiment (like MLM40) you have to make and push a new container. Whereas if the non-default arguments can be passed on from the job script, you would only have to make a new container. This probably still leaves you with some experiments you do have to build a new container for (Bertlarge vs normal maybe?), but for arguments parameterised in the cli of pretrain.py it feels like it should be possible?

Comment (AstroVasseL) [reply asmtransformers/scripts/launch-pretrain.sh:22]: This is difficult, since variables in bash always get passed on as strings. Therefore, anything that is a float of int will be a string until an arithmic operation will be performed on it. That is why I chose to put it here.

File: asmtransformers/pretrain.Containerfile (hunk 4 above)
Comment (AstroVasseL) [reply asmtransformers/pretrain.Containerfile]: I don't mind putting it in the /scripts/ folder, I think one could argue it fits there.

Comment (akaIDIOT) [reply asmtransformers/pretrain.Containerfile]: It's quite literally a script, as there's no external tool that misbehaves when this is not in the root, I'd say put it in scripts? :)

File: asmtransformers/scripts/launch-pretrain.sh (hunk 1 above)
Comment (akaIDIOT) [reply asmtransformers/scripts/launch-pretrain.sh]: And will we include a comment or something with that information here for future reference? :)
